[SwiftUI] Introduce WebPage.Configuration, URLSchemeHandler, and supporting types

rr-codes opened the pull request:
#### 04b52903636aff9541e432dd6474227993572de7
<pre>
[SwiftUI] Introduce WebPage.Configuration, URLSchemeHandler, and supporting types
<a href="https://bugs.webkit.org/show_bug.cgi?id=280598">https://bugs.webkit.org/show_bug.cgi?id=280598</a>
<a href="https://rdar.apple.com/136945136">rdar://136945136</a>

Reviewed by Wenson Hsieh.

Combined changes:
* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _uncheckedSetURLSchemeHandler:forCanonicalURLScheme:]):
(+[WKWebViewConfiguration _canonicalSchemeForURLScheme:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
* Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift: Added.
(URLSchemeHandler_v0.reply(for:)):
(WKURLSchemeHandlerAdapter.tasks):
(WKURLSchemeHandlerAdapter.webView(_:start:)):
(WKURLSchemeHandlerAdapter.webView(_:stop:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift: Added.
(Configuration.websiteDataStore):
(Configuration.userContentController):
(Configuration.webExtensionController):
(Configuration.defaultNavigationPreferences):
(Configuration.urlSchemeHandlers):
(Configuration.applicationNameForUserAgent):
(Configuration.limitsNavigationsToAppBoundDomains):
(Configuration.upgradeKnownHostsToHTTPS):
(Configuration.suppressesIncrementalRendering):
(Configuration.allowsInlinePredictions):
(Configuration.supportsAdaptiveImageGlyph):
(Configuration.dataDetectorTypes):
(Configuration.ignoresViewportScaleLimits):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift: Added.
(preferredContentMode):
(allowsContentJavaScript):
(isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift: Added.
(URL.aboutBlank):
* Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift: Added.
(reply(for:)):
(URLSchemeHandlerTests.basicSchemeValidation):
(URLSchemeHandlerTests.basicSchemeHandling):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(URL.aboutBlank): Deleted.

Canonical link: <a href="https://commits.webkit.org/287406@main">https://commits.webkit.org/287406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47ab155100aa5392d35b92b5aeb45a98c62aae8a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84116 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81671 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67639 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6825 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62199 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20059 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52258 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72464 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42508 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49604 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26626 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29046 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70725 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27084 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85527 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6803 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4741 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70448 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6968 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68325 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69692 "Found 7 new API test failures: /TestWebKit:WebKit.DownloadDecideDestinationCrash, /TestWebKit:WebKit.LoadAlternateHTMLStringWithEmptyBaseURL, /TestWebKit:WebKit.FailedLoad, /TestWebKit:WebKit.WillSendSubmitEvent, /TestWebKit:WebKit.EnumerateDevicesCrash, /TestWebKit:WebKit.ParentFrame, /TestWebKit:WebKit.Find (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13724 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12615 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12286 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6755 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6651 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10131 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8450 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -50,3 +53,13 @@ _WKDragLiftDelay toDragLiftDelay(NSUInteger);
+
+@interface WKWebViewConfiguration (NonCpp)
+
++ (/* nullable */ NSString *)_canonicalSchemeForURLScheme:(NSString *)urlScheme;
```
Why not actually mark this `nullable`?

rr-codes replied:
I tried that originally, but I got a lottt of

```
Pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified)
```

elsewhere in the file

pxlcoder replied:
I think you just need `NS_ASSUME_NONNULL_BEGIN` / `NS_ASSUME_NONNULL_END`?

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -50,3 +53,13 @@ _WKDragLiftDelay toDragLiftDelay(NSUInteger);
+
+@interface WKWebViewConfiguration (NonCpp)
+
++ (nullable NSString *)_canonicalSchemeForURLScheme:(NSString * _Nonnull)urlScheme;
```
```suggestion
+ (nullable NSString *)_canonicalSchemeForURLScheme:(nonnull NSString *)urlScheme;
```
The reason you're being told to put nonnull is because this isn't within an audit region. Auditing the C++ parts above is maybe difficult to do right now, but I think you could probably still add the guards around just this interface.

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -50,3 +53,13 @@ _WKDragLiftDelay toDragLiftDelay(NSUInteger);
+
+#endif // __cplusplus
+
+@interface WKWebViewConfiguration (NonCpp)
```
```suggestion
@interface WKWebViewConfiguration ()
```
You should declare this as a class extension (not a named category). Categories should technically have their implementation supplied in a matching `@implementation Foo (Bar)` block, where you're just implementing it in the primary class implementation, but also there's just no reason for it to be category when you're the defining framework for a type.

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -50,3 +53,13 @@ _WKDragLiftDelay toDragLiftDelay(NSUInteger);
+
++ (nullable NSString *)_canonicalSchemeForURLScheme:(NSString * _Nonnull)urlScheme;
+
+- (void)_uncheckedSetURLSchemeHandler:(id <WKURLSchemeHandler> _Nonnull)urlSchemeHandler forCanonicalURLScheme:(NSString * _Nonnull)urlScheme;
```
Ditto, depending on what you do with the above.

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+public struct URLScheme_v0: Sendable, Hashable, Equatable {
+    let rawValue: String
+
+    public init?(_ rawValue: String) {
```
Hmm. I'm not sure if you should necessarily use the "raw value" terminology if this isn't also `RawRepresentable`. I also don't know if this _should_ be `RawRepresentable` since you transform the input type, so it isn't really the "raw" value anymore.

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+internal import WebKit_Internal
+
+@_spi(Private)
+public struct URLScheme_v0: Sendable, Hashable, Equatable {
```
```suggestion
public struct URLScheme_v0: Sendable, Hashable {
```
You can omit `Equatable` since `Hashable` already refines it. IDK if WebKit does this, but we tend to sort protocols alphabetically. (I sort a lot of stuff just to have _some_ defined order.)

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+public protocol URLSchemeHandler_v0 {
+    associatedtype TaskSequence: AsyncSequence<URLSchemeTaskResult_v0, any Error>
+
+    func reply(for request: URLRequest) -> TaskSequence
```
`TaskSequence` does kinda run afoul of `Swift.Task`. We might want to workshop this a bit more.

I also wonder if we should call this something like
```swift
func handle(_ request: URLRequest) -> Result
```
to lean in to the "handler" part of the name, but probably fine as-is for now

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+// MARK: Adapters
+
+final class WKURLSchemeHandlerAdapter: NSObject, WKURLSchemeHandler {
+    private let wrapped: any URLSchemeHandler_v0
```
Especially because this is `internal` there's no reason to not make it generic over the actual type of the handler.
```swift
final class WKURLSchemeHandlerAdapter<Handler>: NSObject, WKURLSchemeHandler
    where Handler: URLSchemeHandler_v0
{
    init(wrapping handler: Handler) {
        self.handler = handler
    }

    private let handler: Handler
```

rr-codes replied:
I originally had this, but the problem is that this type can never actually be created in this case, since I need to create it using `any URLSchemeHandler` since that's the value type of the populated handler dictionary.

axiixc replied:
Ah, yeah, you’d need a type eraser one way or another then. I guess might as well have this do double duty.

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+            do {
+                for try await result in wrapped.reply(for: urlSchemeTask.request) {
+                    switch result {
+                    case .response(let response):
```
This is a me nit, but I prefer to always put a `let` before the case name, that way it's consistent whether you have 1 or _n_ associated values:
```swift
switch payload {
case .zero: break
case let .one(a): break
case let .two(a, b): break
}
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+                for try await result in wrapped.reply(for: urlSchemeTask.request) {
+                    switch result {
+                    case .response(let response):
+                        urlSchemeTask.didReceive(response)
```
Is it valid to return a mix of `URLResponse` and `Data`s, or is there some sort of defined interleaving here?

rr-codes replied:
according to Brady,

> Much like an HTTP connection:
1 - There can be more than one response (e.g. multiple server redirects)
2 - You can complete without any data (empty response)
3 - There cannot be any data before at least 1 response, but there can be subsequent responses after some data (e.g. multi-part responses)

axiixc replied:
Yeah, that's probably not worth trying to model in the type system.

axiixc commented on Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    func webView(_ webView: WKWebView, stop urlSchemeTask: any WKURLSchemeTask) {
+        tasks.removeValue(forKey: ObjectIdentifier(urlSchemeTask))?.cancel()
```
Something to double check is how the cancellation propagation works here. I'm not sure if it's partially in the client's control, and if so whether we can do things to make it more robust to prevent misuse (ex: checking `Task.isCancelled` on their behalf before processing results maybe?). Don't make changes for this right now though.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift:
```diff
@@ -0,0 +1,52 @@
+    public struct Configuration {
+        public var urlSchemeHandlers: [URLScheme_v0 : any URLSchemeHandler_v0] = [:]
+
+        public init() {
```
I'm still gonna keep voting for `init` before `var` :P

Is there any reason to not accept a dictionary in `init` as well?

rr-codes replied:
> Is there any reason to not accept a dictionary in init as well?

The `Configuration` type is soon going to have around a dozen or so properties, so I figured they should just be mutable `var`s instead of having a giant initializer with a lot of default values

axiixc replied:
Good point. I got tunnel vision here.

axiixc commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -18660,7 +18666,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "if [ \"${ACTION}\" = \"installhdrs\" ] || [ \"${ACTION}\" = \"installapi\" ]; then\n    exit 0;\nfi\n\nif [ -f ../../Tools/Scripts/check-for-inappropriate-objc-class-names ]; then\n    ../../Tools/Scripts/check-for-inappropriate-objc-class-names WK _WK _TtC6WebKit27WK || exit $?\nfi\n";
+			shellScript = "if [ \"${ACTION}\" = \"installhdrs\" ] || [ \"${ACTION}\" = \"installapi\" ]; then\n    exit 0;\nfi\n\nif [ -f ../../Tools/Scripts/check-for-inappropriate-objc-class-names ]; then\n    ../../Tools/Scripts/check-for-inappropriate-objc-class-names WK _WK _TtC6WebKit27WK _TtC6WebKit25WK || exit $?\nfi\n";
```
You want this to be
```suggestion
			shellScript = "if [ \"${ACTION}\" = \"installhdrs\" ] || [ \"${ACTION}\" = \"installapi\" ]; then\n    exit 0;\nfi\n\nif [ -f ../../Tools/Scripts/check-for-inappropriate-objc-class-names ]; then\n    ../../Tools/Scripts/check-for-inappropriate-objc-class-names WK _WK _TtC6WebKit _TtCC6WebKit || exit $?\nfi\n";
```
otherwise you're going to keep having to do this. The `27WK` and `25WK` are referring to the length of the class name to follow, so you'd need like one per every class name length. ^^Those just encode the two possible module prefix mangling of `WebKit` (class and uh… different class thing? I'm not an expert on Swift's name mangling.)

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
```diff
@@ -0,0 +1,44 @@
+import Foundation
+
+extension Array {
+    init(_ sequence: some AsyncSequence<Element, Never>) async {
```
This is a kinda dangerous extension, since it could consume an infinite sequence and never return. Less dangerous being _just_ in the tests, but I wouldn't deploy this in production code.

rr-codes replied:
ya definitely; this is really convenient for testing though since it lets me easily compare an async sequence with an expected sequence

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift:
```diff
@@ -0,0 +1,44 @@
+}
+
+extension URL {
+    static var aboutBlank: URL {
```
Feel free to define short getters like this one line.
```suggestion
    static var aboutBlank: URL { URL(string: "about:blank")! }
```

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,117 @@
+
+import Testing
+import WebKit
+@_spi(Private) import WebKit
```
This also implies the `import WebKit` from above.

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,117 @@
+@_spi(Private) import WebKit
+
+fileprivate struct TestURLSchemeHandler: URLSchemeHandler_v0, Sendable {
+    enum Error: Swift.Error {
```
If you only have one error this could just be
```swift
struct Failure: Error {
}
```

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,117 @@
+        self.data = data
+        self.mimeType = mimeType
+
+        let (stream, continuation) = AsyncStream.makeStream(of: URL.self)
```
I think you can do
```suggestion
        (self.replyStream, self.replyContinuation) = AsyncStream.makeStream(of: URL.self)
```

_Aside: I know I said to avoid explicit `self`, I personally try to use it consistently `init` just because those are usually short and I find it more confusing to mix-and-matching depending only on shadowing, but feel free to ignore me here._

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -23,9 +23,12 @@
 
+#import <WebKit/WKWebViewConfigurationPrivate.h>
+
+#ifdef __cplusplus
```
I think it also _is_ possible to consume C++ in Swift modules now, although that might not be unblocked yet.

rr-codes replied:
C++ interop is currently disabled in WebKit (I forget exactly why, there's a bug tracking it though)

axiixc commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -567,6 +567,23 @@ - (void)setURLSchemeHandler:(id <WKURLSchemeHandler>)urlSchemeHandler forURLSche
 
+- (void)_uncheckedSetURLSchemeHandler:(id<WKURLSchemeHandler>)urlSchemeHandler forCanonicalURLScheme:(NSString *)urlScheme
+{
+    _pageConfiguration->setURLSchemeHandlerForURLScheme(WebKit::WebURLSchemeHandlerCocoa::create(urlSchemeHandler), urlScheme);
```
This could still be checked, if we want more assurance we never register a bad scheme here. Like aside from performance would there be a reason not to double check?

rr-codes replied:
ya performance and redundancy were my main motivation, since this function will only ever be called always after validation

whsieh replied:
Unless this is a really hot codepath, I'd imagine it's better to safer to check here as well. Maybe throw an `NSException` if the scheme is bad?

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift:
```diff
@@ -0,0 +1,97 @@
+// MARK: Adapters
+
+extension WKWebViewConfiguration {
+    convenience init(wrapping wrapped: WebPage_v0.Configuration) {
```
"Wrapping" might be the wrong term here, since it's a one-time copy. TBH I think this could fall under the Swift convention of "converting initializers", which use an unlabeled `init(_:)`.

We could also implement this the other way around, by making `WebPage.Configuration` hold a private reference to a `WKWebViewConfiguration` and doing manual CoW to preserve value semantics. That would let you just grab out the `WebPage.Configuration().wkConfiguration` internally (though implementing CoW is itself non-trivial ¯\\_(ツ)\_/¯)

axiixc replied:
This is more speculative, but you might even be able to use dynamic member lookup with that to auto-synthesize all the getters. The catch is that we might want to hide or refine certain properties, so we'd want to double check on whether we can reserve that kind of flexibility, or whether some other form of more static code generation would be better then.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -0,0 +1,76 @@
+    @MainActor
+    @_spi(Private)
+    public struct NavigationPreferences: Sendable {
+        public enum ContentMode: Sendable {
```
Could we reuse the existing enum here? It doesn't seem particularly different.

rr-codes replied:
I did this because the existing enum is `WKWebpagePreferences.ContentMode`, which felt weird to use within a type named `NavigationPreferences`

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -0,0 +1,76 @@
+
+        public var allowsContentJavaScript: Bool = true
+
+        public var isLockdownModeEnabled: Bool = false {
```
You could also model this with an internal `Bool?`:
```swift
fileprivate var _isLockdownModeEnabled: Bool? = nil
public var isLockdownModeEnabled: Bool {
    get { _isLockdownModeEnabled ?? false }
    set { _isLockdownModeEnabled = newValue }
}
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -0,0 +1,76 @@
+    convenience init(wrapping wrapped: WebPage_v0.NavigationPreferences) {
+        self.init()
+
+        self.preferredContentMode = switch wrapped.preferredContentMode {
```
This conversion seems better defined in an extension, or by just reusing the existing enum.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift:
```diff
@@ -0,0 +1,76 @@
+
+        self.allowsContentJavaScript = wrapped.allowsContentJavaScript
+
+        if wrapped.didSetIsLockdownModeEnabled {
```
This could then be:
```swift
if let isLockdownModeEnabled = wrapped._isLockdownModeEnabled {
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -101,12 +103,14 @@ public class WebPage_v0 {
     @ObservationIgnored
     lazy var backingWebView: WKWebView = {
-        let webView = WKWebView(frame: .zero)
+        let webView = WKWebView(frame: .zero, configuration: WKWebViewConfiguration(wrapping: configuration))
```
Wrapping would also be nice because it'd mean that `public let configuration: Configuration` could then be inverted to return like `{ Configuration(storage: backingWebView.configuration) }` which just helps a little bit with canonicalization.

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,113 @@
+@_spi(Private) import WebKit
+
+fileprivate struct TestURLSchemeHandler: URLSchemeHandler_v0, Sendable {
+    struct Error: Swift.Error {
```
I called my example `Failure`, because if you do this I can bet that you (or someone else) is gonna write something like `let error: Error` and have a very confusing compiler error in the future. Also so you don't need to write out `any Swift.Error` below :P If you don't like `Failure` maybe `ExampleError`, `TestError`, `MockError`?

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift:
```diff
@@ -0,0 +1,113 @@
+
+        let expectedReplyURLs = [
+            URL(string: "testing:main")!,
+            URL(string: "testing:image")!
```
```suggestion
            URL(string: "testing:image")!,
```

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
```diff
@@ -41,7 +44,7 @@ template<> struct WrapperTraits<API::PageConfiguration> {
 }
 
-@property (nonatomic, readonly) NSString *_applicationNameForDesktopUserAgent;
+@property (nonatomic, readonly, nullable) NSString *_applicationNameForDesktopUserAgent;
```
Are the nullability annotations required? I'm a bit surprised this doesn't bleed into the other internal headers when they include WKWebViewConfigurationInternal.h.

rr-codes replied:
yes, I get an error when I exclude it

rr-codes committed:
[SwiftUI] Introduce WebPage.Configuration, URLSchemeHandler, and supporting types
https://bugs.webkit.org/show_bug.cgi?id=280598
rdar://136945136

Reviewed by Wenson Hsieh.

Combined changes:
* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _uncheckedSetURLSchemeHandler:forCanonicalURLScheme:]):
(+[WKWebViewConfiguration _canonicalSchemeForURLScheme:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationInternal.h:
* Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift: Added.
(URLSchemeHandler_v0.reply(for:)):
(WKURLSchemeHandlerAdapter.tasks):
(WKURLSchemeHandlerAdapter.webView(_:start:)):
(WKURLSchemeHandlerAdapter.webView(_:stop:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift: Added.
(Configuration.websiteDataStore):
(Configuration.userContentController):
(Configuration.webExtensionController):
(Configuration.defaultNavigationPreferences):
(Configuration.urlSchemeHandlers):
(Configuration.applicationNameForUserAgent):
(Configuration.limitsNavigationsToAppBoundDomains):
(Configuration.upgradeKnownHostsToHTTPS):
(Configuration.suppressesIncrementalRendering):
(Configuration.allowsInlinePredictions):
(Configuration.supportsAdaptiveImageGlyph):
(Configuration.dataDetectorTypes):
(Configuration.ignoresViewportScaleLimits):
* Source/WebKit/UIProcess/API/Swift/WebPage+NavigationPreferences.swift: Added.
(preferredContentMode):
(allowsContentJavaScript):
(isLockdownModeEnabled):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/Foundation+Extras.swift: Added.
(URL.aboutBlank):
* Tools/TestWebKitAPI/Tests/WebKit Swift/URLSchemeHandler.swift: Added.
(reply(for:)):
(URLSchemeHandlerTests.basicSchemeValidation):
(URLSchemeHandlerTests.basicSchemeHandling):
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
(URL.aboutBlank): Deleted.

Canonical link: https://commits.webkit.org/287406@main
